chasm/field.go

```diff
@@ -47,21 +49,64 @@ type Field[T any] struct {
 }
 
 type fieldInternal struct {
-	fieldType fieldType
-	value     any // Component | Data | Pointer
+	// These 2 fields are used when node is not created yet.
+	// Don't access them directly outside of this file. Use corresponding getters instead.
+	ft fieldType
+	v  any // Component | Data | Pointer
 
-	// Pointer to the corresponding tree node. Can be nil for the just assigned fields.
+	// Pointer to the corresponding tree node. Can be nil for the just created fields.
 	node *Node
 }
 
-func (fi fieldInternal) IsEmpty() bool {
-	return fi.value == nil
+func (fi fieldInternal) isEmpty() bool {
+	return fi.value() == nil
 }
 
-func (f Field[T]) Get(Context) (T, error) {
-	// remember to handle d == nil case
+func (fi fieldInternal) value() any {
+	if fi.node == nil {
+		return fi.v
+	}
+	return fi.node.value
+}
 
-	panic("not implemented")
+func (fi fieldInternal) fieldType() fieldType {
+	if fi.node == nil {
+		return fi.ft
+	}
+	return fi.node.fieldType()
+}
+
+func (f Field[T]) Get(chasmContext Context) (T, error) {
+	// If value is not nil, use it.
+	if f.Internal.v != nil {
+		return f.Internal.v.(T), nil
+	}
+
+	// If node is nil, then there is nothing to deserialize from, return nil.
+	if f.Internal.node == nil {
+		var nilT T
+		return nilT, nil
+	}
+
+	var err error
+	switch f.Internal.fieldType() {
+	case fieldTypeComponent:
+		err = f.Internal.node.prepareComponentValue(chasmContext)
+	case fieldTypeData:
+		// For data fields, T is always a concrete type.
+		err = f.Internal.node.prepareDataValue(chasmContext, reflect.TypeFor[T]())
+	case fieldTypeComponentPointer:
+		panic("not implemented")
+	default:
+		err = serviceerror.NewInternal(fmt.Sprintf("unsupported field type: %v", f.Internal.fieldType()))
+	}
+
+	if err != nil {
+		var nilT T
+		return nilT, err
+	}
+	// TODO: check if f.Internal.node.value is T or implements T?
```

Comment:
yeah it would be nice to have the check and return a proper error message and not panic.

Reply:
linter won`t pass this anyway.